lib/grandpa/finalization.go

```diff
@@ -0,0 +1,493 @@
+				}
+
+				// broadcast pre-vote
+				preVote, err := h.grandpaService.determinePreVote()
+				if err != nil {
+					return fmt.Errorf("determining pre-vote: %w", err)
+				}
+
+				signedpreVote, prevoteMessage, err :=
+					h.grandpaService.createSignedVoteAndVoteMessage(preVote, prevote)
+				if err != nil {
+					return fmt.Errorf("creating signed vote: %w", err)
+				}
+
+				if !isPrimary {
+					h.grandpaService.prevotes.Store(h.grandpaService.publicKeyBytes(), signedpreVote)
+				}
+
+				logger.Warnf("sending pre-vote message: {%v}", prevoteMessage)
+				h.grandpaService.sendPrevoteMessage(prevoteMessage)
+
+			case determinePrecommit:
+				preComit, err := h.grandpaService.determinePreCommit()
+				if err != nil {
+					return fmt.Errorf("determining pre-commit: %w", err)
+				}
+
+				signedpreComit, precommitMessage, err :=
+					h.grandpaService.createSignedVoteAndVoteMessage(preComit, precommit)
+				if err != nil {
+					return fmt.Errorf("creating signed vote: %w", err)
+				}
+
+				logger.Warnf("sending pre-commit message: {%v}", precommitMessage)
+				h.grandpaService.precommits.Store(h.grandpaService.publicKeyBytes(), signedpreComit)
+				h.grandpaService.sendPrecommitMessage(precommitMessage)
+
+			case finalize:
+				commitMessage, err := h.grandpaService.newCommitMessage(
+					h.grandpaService.head, h.grandpaService.state.round, h.grandpaService.state.setID)
+				if err != nil {
+					return fmt.Errorf("creating commit message: %w", err)
+				}
+
+				commitConsensusMessage, err := commitMessage.ToConsensusMessage()
+				if err != nil {
+					return fmt.Errorf("transforming commit into consensus message: %w", err)
+				}
+
+				logger.Debugf("sending commit message: %v", commitMessage)
+
+				h.grandpaService.network.GossipMessage(commitConsensusMessage)
+				h.grandpaService.telemetry.SendMessage(telemetry.NewAfgFinalizedBlocksUpTo(
+					h.grandpaService.head.Hash(),
+					fmt.Sprint(h.grandpaService.head.Number),
+				))
+
+				logger.Debugf("round completed in %s", time.Since(start))
+				return nil
+
+			case alreadyFinalized:
+				logger.Debugf("round completed in %s", time.Since(start))
+				return nil
+			}
+		}
+	}
+}
+
+// actions that should take place accordingly to votes the
+// finalisation engine knows about
+type engineAction byte
+
+const (
+	determinePrevote engineAction = iota
+	determinePrecommit
+	alreadyFinalized
+	finalize
+)
+
+type finalizationEngine struct {
+	grandpaService *Service
+
+	timeoutStop time.Duration
+	stopCh      chan struct{}
+	engineDone  chan struct{}
+	actionCh    chan engineAction
+}
+
+func newFinalizationEngine(service *Service) *finalizationEngine {
+	return &finalizationEngine{
+		grandpaService: service,
+		timeoutStop:    5 * time.Second,
+		actionCh:       make(chan engineAction),
+		stopCh:         make(chan struct{}),
+		engineDone:     make(chan struct{}),
+	}
+}
+
+func (f *finalizationEngine) Stop() (err error) {
+	if f.stopCh == nil {
+		return nil
+	}
+
```

Comment:
yes, since the code is designed to call `stop` when a service fails it is possible that we call the `Stop()` method from a service that is already stopped